### ISSUE TITLE
Slow queries when exporting Edges

### ISSUE THREAD

**Issue by frgao9** [aws/neptune-export#57]
I'm using AWS Neptune Export Service to export graph into .csv files, and it trigger neptune-export using cmd "export-pg".
I observed whenever exporting Edges with "gremlinFilter", it first counts all Edges in graph, and range the query based on concurrency, then apply the filter on ranged queries, which makes the query and exporting very slow.

see below logs - 
why is the filter applied on ranges ```__.E().range(85979418L,90073676L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2")```
instead of range on filter ```__.E().has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").range(...)```, would this be much faster?

I saw it is doing the other way around (filter then range) when exporting Nodes, is there a specific reason why not query the same way as Nodes when exporting Edges?

```
Completed counting all edges in 335 seconds
[main] INFO com.amazonaws.services.neptune.propertygraph.RangeFactory - Estimated number of all edges to export: 393048717, Range size: 4094258, Effective concurrency: 96
Started writing all edges as CSV to files
[pool-5-thread-22] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(85979418L,90073676L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-23] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(90073676L,94167934L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-42] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(167864578L,171958836L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-32] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(126921998L,131016256L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-54] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(216995674L,221089932L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-63] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(253843996L,257938254L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-65] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(262032512L,266126770L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-58] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(233372706L,237466964L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-81] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(323446382L,327540640L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
[pool-5-thread-3] INFO com.amazonaws.services.neptune.propertygraph.EdgesClient - __.E().range(32754064L,36848322L).has("_partition","47109520-440a-41e7-805a-84b1c1ff86b2").project("~id","~label","properties","~from","~to").by(T.id).by(T.label).by(__.valueMap()).by(__.outV().id()).by(__.inV().id())
...
```

**Comment by Cole-Greer** [comment]
Hi @frgao9, I'm not aware of the initial motivation behind placing the range before the filter for edges, however I can certainly run some tests to determine if there is any benefit to switching this order.

I'm unsure which order I would intuitively expect to be fastest. If the filter is slow to evaluate, I would expect the early range() to be the best choice to limit the input size to the filter. However if the filter is fast I would agree with you in expecting filtering first to be best.

I will run some tests regarding this and report back any findings I have.

**Comment by frgao9** [comment]
Thank you @Cole-Greer for investigating, i believe for most case it would be much faster to run filter first, and users could also decide whether to send "range()" in gremlinFilter if needed, because they know better of their data. 
For my case, I have hundreds of millions of nodes and edges in database. It only take seconds with less cpu and memory to export nodes when using filter, however hours to export edges even the most powerful cpu and memory since it search entire database by "range".

**Comment by frgao9** [comment]
just to follow up, any update for this?

**Comment by Cole-Greer** [comment]
Hi @frgao9, Sorry for the late reply, I have been away for a little while. I've been running some experiments with this and so far the results have been poor. I'm running tests with a distribution of filters (range from filtering most edges to very few) and so far I've only observed significantly worse performance when applying the range after the filter. Additionally I am noting that the edge counting phase of the export can become incredibly slow depending on the filter as the graph may need to apply the filter to all edges first in order to count them.

I have a few more tests I would like to run and then I can share some more quantitative results.

**Comment by Cole-Greer** [comment]
@frgao9 My tests are showing that the results are very dependent on the specific "gremlin-filter" being applied. If it's a simple filter or something which can be processed very efficiently by Neptune, pushing the filter ahead of the range does indeed give a bump to performance. The bump in performance is most substantial for filters which remove the vast majority of the edges from the traversal. The best case I measured was a 50% reduction in edge writing time. However if the filter is something more complex and slow to process, pushing it down before the range can lead to substantially worse performance. The worst I've observed is the edge writing time more than double with the change. 

The other factor here is that if the filter gets applied before the range, the filter must also be applied in the edge counting phase of the export. Depending on the graph and filter, this may increase the count time from a few seconds to several minutes.

Based on these findings, I wouldn't want to simply replace the current behavior with a default "early-filter" variant. Would you find it acceptable if we were to add a new configuration option such as `--filter-edges-early` which would pull the filters ahead of the ranges?

**Comment by frgao9** [comment]
@Cole-Greer, you are right, the exporting performance is various depends on how filter( as well as data) looks like.
And you suggestion with argument "--filter-edges-early" make sense to me, thank you for all investigations and proposed solutions!

**Comment by Cole-Greer** [comment]
Thanks @frgao9, our next release cycle is at the end of the month. I expect we will have the new option included in that release.

**Comment by Cole-Greer** [comment]
The new `--filter-edges-early` now merged in and will be included in the upcoming v1.1.0 which is expected to be released tomorrow.
